src/hamiltonian.jl

```diff
@@ -15,24 +16,53 @@ end
 ∂H∂r(h::Hamiltonian{<:DiagEuclideanMetric}, r::AbstractVector) = h.metric.M⁻¹ .* r
 ∂H∂r(h::Hamiltonian{<:DenseEuclideanMetric}, r::AbstractVector) = h.metric.M⁻¹ * r
 
-function hamiltonian_energy(h::Hamiltonian, θ::AbstractVector, r::AbstractVector)
-    K = kinetic_energy(h, r, θ)
-    if isnan(K)
-        K = Inf
-        @warn "Kinetic energy is `NaN` and is set to `Inf`."
-    end
-    V = potential_energy(h, θ)
-    if isnan(V)
-        V = Inf
-        @warn "Potential energy is `NaN` and is set to `Inf`."
+
+# The constructor of `DualValue` will check numerical errors in
+#   `value` and `gradient`.  That is `is_valid` will be performed automatically.
+struct DualValue{Tv<:AbstractFloat, Tg<:AbstractVector{Tv}} <: AbstractFloat
+    value::Tv    # Cached value, e.g. logπ(θ).
+    gradient::Tg # Cached gradient, e.g. ∇logπ(θ).
+end
+
+# TODO: replace logπ and logκ with ℓπ, ℓκ??
+# The constructor of `PhasePoint` will check numerical errors in
+#   `θ`, `r` and `h`. That is `is_valid` will be performed automatically.
+struct PhasePoint{T<:AbstractVector, V<:AbstractFloat}
+    θ::T # position variables / parameters
+    r::T # momentum variables
+    logπ::V # cached potential energy for the current θ
+    logκ::V # cached kinect energy for the current r
+    function PhasePoint(θ::T, r::T, ℓπ::V, ℓκ::V) where {T,V}
+        @argcheck length(θ) == length(r) == length(ℓπ.gradient) == length(ℓπ.gradient)
+        # if !(all(isfinite, θ) && all(isfinite, r) && all(isfinite, ℓπ) && all(isfinite, ℓκ))
+        if !(isfinite(θ) && isfinite(r) && isfinite(ℓπ) && isfinite(ℓκ))
+            @warn "A numerical error is detected. The current proposal will be rejected..."
+            ℓκ = DualValue(-Inf, ℓκ.gradient)
+            ℓπ = DualValue(-Inf, ℓπ.gradient)
+        end
+        new{T,V}(θ, r, ℓπ, ℓκ)
     end
-    return K + V
 end
 
+function phasepoint(h::Hamiltonian, θ::AbstractVector, r::AbstractVector)
+    π = DualValue(-kinetic_energy(h, r, θ), ∂H∂θ(h, θ))
+    κ = DualValue(-potential_energy(h, θ), ∂H∂r(h, r))
+    return PhasePoint(θ, r, π, κ)
+end
+
+Base.isfinite(v::DualValue) = all(isfinite, v.value) && all(isfinite, v.gradient)
+Base.isfinite(v::AbstractVector) = all(isfinite, v)
+Base.isfinite(z::PhasePoint) = isfinite(z.logπ) && isfinite(z.logκ)
+
+###
+### Negative energy (or log probability) functions.
+###
+
+neg_energy(z::PhasePoint) = - z.logπ.value - z.logκ.value
```

Comment:
Why do we change the name here? I think (potential) energy usually refers to `-z.logπ.value` only (as what is is said in the comments above but what this computes is the Hamiltonian. Maybe we could call it "total_energy" if you don't like the old "hamiltonian_energy".

Reply:
I plan to unify the interface for all energy functions (e.g. potential_energy, kinetic_energy) to `negative_energy`. This requires some further work, e.g. creating a type for `kinect` energy.

Reply:
I changed all `*_energy` functions to `neg_energy`. Please let me know if this works for you.

Reply:
So the design is intended to do
`neg_energy(z::PhasePoint)` => potential + kinetic
`neg_energy(z::Potential)` => potential energy
`neg_energy(z::Kinetic)` => kinetic energy
?

Sound like a good design to me!

Reply:
yes